lib/matplotlib/tests/test_axes.py

```diff
@@ -5047,3 +5047,145 @@ def test_invalid_axis_limits():
+    'Test longer legend with set order.'
+    x = np.linspace (0,50,400)
+    fig,ax = plt.subplots(1)
+    ax.plot(x, x + 1, lorder=1, label="Test 1")
+    ax.plot(x, x - 1, lorder=4, label="Test 4")
+    ax.plot(x, x - 1, lorder=5, label="Test 5")
+    ax.plot(x, x - 1, lorder=2, label="Test 2")
+    ax.plot(x, x - 1, lorder=3, label="Test 3")
+
+    handles = ['Line2D(Test 1)', 'Line2D(Test 2)', 'Line2D(Test 3)',
+               'Line2D(Test 4)', 'Line2D(Test 5)']
+    labels = [u'Test 1', u'Test 2', u'Test 3', u'Test 4', u'Test 5']
+
+    legend = plt.legend()
+
+    handles_equal = [(str(handle1) == str(handle2)) for (handle1, handle2) in
+                     zip(legend.legendHandles, handles)]
+    handles_equal = reduce((lambda x, y: x and y), handles_equal)
+    assert handles_equal
+
+    texts = [x.get_text() for x in legend.texts]
+    texts_equal = texts == labels
+    assert texts_equal
+
+
+def test_legend_lorder_duplicate_values():
+    '''Test longer legends behaviour when two lorder kwargs 
+    have the same value.'''
+    x = np.linspace (0,50,400)
+    fig,ax = plt.subplots(1)
+    ax.plot(x, x + 1, lorder=1, label="Test 1")
+    ax.plot(x, x - 1, lorder=4, label="Test 5")
+    ax.plot(x, x - 1, lorder=2, label="Test 2")
+    ax.plot(x, x - 1, lorder=2, label="Test 3")
+    ax.plot(x, x - 1, lorder=3, label="Test 4")
+
+    handles = ['Line2D(Test 1)', 'Line2D(Test 2)', 'Line2D(Test 3)',
+               'Line2D(Test 4)', 'Line2D(Test 5)'] 
+    labels = [u'Test 1', u'Test 2', u'Test 3', u'Test 4', u'Test 5']
+
+    legend = plt.legend()
+
+    handles_equal = [(str(handle1) == str(handle2)) for (handle1, handle2) in
+                     zip(legend.legendHandles, handles)]
+    handles_equal = reduce((lambda x, y: x and y), handles_equal)
+    assert handles_equal
+
+    texts = [x.get_text() for x in legend.texts]
+    texts_equal = texts == labels
+    assert texts_equal
+
+
+def test_legend_lorder_random_values():
+    'Test longer legends maintains relative order with random values.'
+    x = np.linspace (0,50,400)
+    fig,ax = plt.subplots(1)
+    ax.plot(x, x + 1, lorder=102, label="Test 3")
+    ax.plot(x, x - 1, lorder=42, label="Test 2")
+    ax.plot(x, x - 1, lorder=2569, label="Test 5")
+    ax.plot(x, x - 1, lorder=0, label="Test 1")
+    ax.plot(x, x - 1, lorder=300, label="Test 4")
+
+    handles = ['Line2D(Test 1)', 'Line2D(Test 2)', 'Line2D(Test 3)',
+               'Line2D(Test 4)', 'Line2D(Test 5)'] 
+    labels = [u'Test 1', u'Test 2', u'Test 3', u'Test 4', u'Test 5']
+
+    legend = plt.legend()
+
+    handles_equal = [(str(handle1) == str(handle2)) for (handle1, handle2) in
+                     zip(legend.legendHandles, handles)]
+    handles_equal = reduce((lambda x, y: x and y), handles_equal)
+    assert handles_equal
+
+    texts = [x.get_text() for x in legend.texts]
+    texts_equal = texts == labels
+    assert texts_equal
+
+
+def test_legend_lorder_some_unordered():
```

Comment:
I'm noticing a lot of code duplication in your tests. 

I think it'd be wise to use pytest to parametrize your tests. Something like:

```python
@pytest.mark.parametrize(('lorder_list', 'label_list'), [
    ([None, None, 1, 2, 3], ['L4', 'L5', 'L1', 'L2', 'L3']),
    ([102, 42, 2569, 0, 380], ['L3', 'L2', 'L5', 'L1', 'L4']),
    ([1, 4, 2, 2, 3], ['L1', 'L5', 'L2', 'L3', 'L4']),
    ([1, 4, 5, 2, 3], ['L1', 'L4', 'L5', 'L2', 'L3']),
])
def test_legend_lorder(lorder_list, label_list):
    x = np.linspace(0, 50, 400)
    fig, ax = plt.subplots()
    for n, (lorder, label) in enumerate(zip(lorder_list, label_list)):
        ax.plot(x, x + 1 - n, lorder=lorder, label=label)
        leg = ax.legend()
        handles = ['Line2D({})'.format(label) for label in label_list]
        assert handles == leg.legenedHandles
        assert label_list == [x.get_text() for x in leg.texts]
```

And that should cover all of your bases